### ISSUE TITLE
Results of Conv2d on GPU without CuDNN support substantially differ from CuDNN/CPU

### ISSUE THREAD

**Issue by DLtangsan** [pytorch/pytorch#55381]
## 🐛 Bug
The final result of running `conv2d` on GPU without CuDNN support (ATen) differs from that run on CPU or GPU with CuDNN support. Specifically, `stride != 1` seems to be the key to the issue because the following script doesn't reveal the bug if `stride = 1`.

## To Reproduce
The main script is as follows
```
import torch
import torch.nn as nn

torch.manual_seed(0)

in_channels = 64
out_channels = 128
stride = 2
W = 32
H = 32

image = torch.rand(1, in_channels, W, H)
net = nn.Conv2d(in_channels, out_channels, kernel_size=1, stride=stride, bias=False)
torch.save(net, '.\model.pkl')

def runcpu(image):
    net = torch.load('.\model.pkl')
    net.eval()
    with torch.no_grad():        
        output = net(image)
    return output

def runCuDNN(image):
    net = torch.load('.\model.pkl').cuda()
    net.eval()
    with torch.no_grad():        
        torch.backends.cudnn.enabled = True
        output = net(image.cuda())
    return output

def runNoCuDNN(image):
    net = torch.load('.\model.pkl').cuda()
    net.eval()
    with torch.no_grad():        
        torch.backends.cudnn.enabled = False
        output = net(image.cuda())
    return output


noCuDNN_output = runNoCuDNN(image)
cpu_output = runcpu(image)
CuDNN_output = runCuDNN(image)
print(cpu_output.sum())
print(CuDNN_output.sum())
print(noCuDNN_output.sum())
```
Here I use `torch.backends.cudnn.enabled = False` to disable CuDNN.
The result is
```
tensor(502.4232)
tensor(502.4232, device='cuda:0')
tensor(541.0007, device='cuda:0')
```

## Expected behavior
The output shows that the results of `runcpu` and `runCuDNN` are consistent, while `runNoCuDNN` is greatly different from them. It is expected that different running backends should produce exactly the same results. Note that `stride != 1` seems to be the key to the issue because the above script doesn't reveal the bug if `stride = 1`.

## Environment
PyTorch version: 1.8.1+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: CentOS Linux release 7.7.1908 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
Clang version: Could not collect
CMake version: version 2.8.12.2

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Tesla V100-PCIE-16GB

Nvidia driver version: 455.45.01
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.7.5.0
/usr/lib64/libcudnn.so.8.0.5
/usr/lib64/libcudnn_adv_infer.so.8.0.5
/usr/lib64/libcudnn_adv_train.so.8.0.5
/usr/lib64/libcudnn_cnn_infer.so.8.0.5
/usr/lib64/libcudnn_cnn_train.so.8.0.5
/usr/lib64/libcudnn_ops_infer.so.8.0.5
/usr/lib64/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.8.1+cu111
[pip3] torchaudio==0.8.1
[pip3] torchvision==0.9.1+cu111
[conda] Could not collect

cc @ngimel @csarofeen @ptrblck @xwang233

**Comment by DLtangsan** [comment]
Specifically, if `stride != 1` holds for `conv2d`, the results will differ.
```
import torch
import torch.nn as nn

torch.manual_seed(0)

in_channels = 64
out_channels = 128
stride = 2
W = 32
H = 32

image = torch.rand(1, in_channels, W, H)
net = nn.Conv2d(in_channels, out_channels, kernel_size=1, stride=stride, bias=False)
torch.save(net, '.\model.pkl')

def runcpu(image):    
    in_channels = 64
    net = torch.load('.\model.pkl')
    net.eval()
    with torch.no_grad():        
        output = net(image)
    return output

def runCuDNN(image):
    net = torch.load('.\model.pkl').cuda()
    net.eval()
    with torch.no_grad():        
        torch.backends.cudnn.enabled = True
        output = net(image.cuda())
    return output

def runNoCuDNN(image):
  
    net = torch.load('.\model.pkl').cuda()
    net.eval()
    with torch.no_grad():        
        torch.backends.cudnn.enabled = False
        output = net(image.cuda())
    return output


noCuDNN_output = runNoCuDNN(image)
cpu_output = runcpu(image)
CuDNN_output = runCuDNN(image)
print(cpu_output.sum())
print(CuDNN_output.sum())
print(noCuDNN_output.sum())
```
The output is
```
tensor(502.4232)
tensor(502.4232, device='cuda:0')
tensor(541.0007, device='cuda:0')
```

**Comment by DLtangsan** [comment]
Can anyone help me confirm this bug? Thanks a lot!
cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @VitalyFedyunin

**Comment by zou3519** [comment]
I think we've fixed this: https://github.com/pytorch/pytorch/issues/54036. Could you try a nightly build of PyTorch, @DLtangsan?

**Comment by ngimel** [comment]
I can't repro on master, so as @zou3519 says, it looks like it's fixed in #54036. Please reopen if nightly doesn't work for you.

**Comment by DLtangsan** [comment]
Yeah, you are right. Nightly has been fixed already. Thanks a lot!
